File: codes/javascript/gpt/partial/task_2/Solution.js
function reallocateBed(patient, hospitals) {
    function matchesRequirements(patient, ward) {
        if (ward.gender_restriction !== "No Restriction" && ward.gender_restriction !== patient.gender) {
            return false;
        }
        if (patient.age < ward.age_restriction[0] || patient.age > ward.age_restriction[1]) {
            return false;
        }
        for (let req of patient.special_requirements) {
            if (!ward.special_requirements.includes(req)) {
                return false;
            }
        }
        return true;
    }

    function findAvailableBed(hospitals, patient, modelPriority) {
        for (let hospital of hospitals) {
            if (modelPriority.includes(hospital.model)) {
                for (let ward of hospital.wards) {
                    if (matchesRequirements(patient, ward)) {
                        if (ward.available_beds.length > 0) {
                            return {
                                assigned_hospital: hospital.id,
                                assigned_ward: ward.ward_name,
                                assigned_bed: ward.available_beds.shift(),
                                reallocated_patients: []
                            };
                        }
                    }
                }
            }
        }
        return null;
    }

    function tryReallocateHighRiskPatient(patient, hospitals) {
        for (let hospital of hospitals) {
            if (hospital.model === 4) {
                for (let ward of hospital.wards) {
                    if (matchesRequirements(patient, ward)) {
                        for (let currentPatient of ward.current_patients) {
                            if (currentPatient.days_in_hospital > 3 && !currentPatient.non_transferable) {
                                let newWard = findAvailableBed(hospitals.filter(h => h.model < 4), currentPatient, [1, 2, 3]);
                                if (newWard) {
                                    ward.current_patients = ward.current_patients.filter(p => p.id !== currentPatient.id);
                                    newWard.reallocated_patients.push({
                                        patient_id: currentPatient.id,
                                        new_hospital: newWard.assigned_hospital,
                                        new_ward: newWard.assigned_ward,
                                        new_bed: newWard.assigned_bed
                                    });
                                    return {
                                        assigned_hospital: hospital.id,
                                        assigned_ward: ward.ward_name,
                                        assigned_bed: currentPatient.bed_number,
                                        reallocated_patients: newWard.reallocated_patients
                                    };
                                }
                            }
                        }
                    }
                }
            }
        }
        return null;
    }

    let modelPriority = patient.is_post_surgery ? [4] : [1, 2, 3, 4];
    let directAssignment = findAvailableBed(hospitals, patient, modelPriority);

    if (directAssignment) {
        return {
            patient_id: patient.id,
            ...directAssignment
        };
    }

    if (patient.is_post_surgery) {
        let reallocationResult = tryReallocateHighRiskPatient(patient, hospitals);
        if (reallocationResult) {
            return {
                patient_id: patient.id,
                ...reallocationResult
            };
        }
    }

    return {
        patient_id: patient.id,
        assigned_hospital: null,
        assigned_ward: null,
        assigned_bed: null,
        reallocated_patients: []
    };
}

const fs = require('fs');
const input = JSON.parse(fs.readFileSync(0, 'utf-8'));
const result = reallocateBed(input.patient, input.hospitals);
console.log(JSON.stringify(result));

File: codes/go/gpt/partial/task_2/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func reallocateBed(patient map[string]interface{}, hospitals []interface{}) map[string]interface{} {
	result := map[string]interface{}{
		"patient_id":         patient["id"],
		"assigned_hospital":  nil,
		"assigned_ward":      nil,
		"assigned_bed":       nil,
		"reallocated_patients": []map[string]interface{}{},
	}

	isHighRisk := patient["is_post_surgery"].(bool)

	var preferredHospitals []interface{}
	if isHighRisk {
		for _, h := range hospitals {
			hospital := h.(map[string]interface{})
			if hospital["model"].(float64) == 4 {
				preferredHospitals = append(preferredHospitals, hospital)
			}
		}
	} else {
		for _, h := range hospitals {
			hospital := h.(map[string]interface{})
			if hospital["model"].(float64) < 4 {
				preferredHospitals = append(preferredHospitals, hospital)
			}
		}
		if len(preferredHospitals) == 0 {
			for _, h := range hospitals {
				hospital := h.(map[string]interface{})
				if hospital["model"].(float64) == 4 {
					preferredHospitals = append(preferredHospitals, hospital)
				}
			}
		}
	}

	for _, h := range preferredHospitals {
		hospital := h.(map[string]interface{})
		for _, w := range hospital["wards"].([]interface{}) {
			ward := w.(map[string]interface{})
			if isWardSuitable(patient, ward) {
				if len(ward["available_beds"].([]interface{})) > 0 {
					assignedBed := ward["available_beds"].([]interface{})[0].(float64)
					result["assigned_hospital"] = hospital["id"]
					result["assigned_ward"] = ward["ward_name"]
					result["assigned_bed"] = assignedBed
					return result
				}
			}
		}
	}

	if isHighRisk {
		for _, h := range preferredHospitals {
			hospital := h.(map[string]interface{})
			for _, w := range hospital["wards"].([]interface{}) {
				ward := w.(map[string]interface{})
				if isWardSuitable(patient, ward) {
					for _, cp := range ward["current_patients"].([]interface{}) {
						currentPatient := cp.(map[string]interface{})
						if currentPatient["days_in_hospital"].(float64) > 3 && !currentPatient["non_transferable"].(bool) {
							for _, lh := range hospitals {
								lHospital := lh.(map[string]interface{})
								if lHospital["model"].(float64) < 4 {
									for _, lw := range lHospital["wards"].([]interface{}) {
										lWard := lw.(map[string]interface{})
										if isWardSuitable(currentPatient, lWard) {
											if len(lWard["available_beds"].([]interface{})) > 0 {
												assignedBed := currentPatient["bed_number"]
												result["assigned_hospital"] = hospital["id"]
												result["assigned_ward"] = ward["ward_name"]
												result["assigned_bed"] = assignedBed
												newBed := lWard["available_beds"].([]interface{})[0].(float64)

												result["reallocated_patients"] = append(result["reallocated_patients"].([]map[string]interface{}), map[string]interface{}{
													"patient_id":  currentPatient["id"],
													"new_hospital": lHospital["id"],
													"new_ward":     lWard["ward_name"],
													"new_bed":      newBed,
												})
												return result
											}
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}

	return result
}

func isWardSuitable(patient map[string]interface{}, ward map[string]interface{}) bool {
	if ward["gender_restriction"] != "No Restriction" && ward["gender_restriction"] != patient["gender"] {
		return false
	}
	age := patient["age"].(float64)
	ageRestriction := ward["age_restriction"].([]interface{})
	if age < ageRestriction[0].(float64) || age > ageRestriction[1].(float64) {
		return false
	}
	for _, req := range patient["special_requirements"].([]interface{}) {
		found := false
		for _, wReq := range ward["special_requirements"].([]interface{}) {
			if req == wReq {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

func main() {
	var input struct {
		Patient   map[string]interface{} `json:"patient"`
		Hospitals []interface{}          `json:"hospitals"`
	}

	decoder := json.NewDecoder(os.Stdin)
	if err := decoder.Decode(&input); err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	result := reallocateBed(input.Patient, input.Hospitals)

	output, err := json.Marshal(result)
	if err != nil {
		fmt.Println("Error encoding output:", err)
		return
	}

	fmt.Println(string(output))
}